### PR TITLE
Feat: generic file system kv store

### PR DESCRIPTION
**Pull request by bracesproul**

### CODE REVIEW

File: langchain/src/storage/file_system.ts [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+      }
+      values = JSON.parse(fileContent);
+      // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    } catch (e: any) {
+      // If the file does not exist, create it
+      if (("code" in e && e.code === "EISDIR") || e.code === "ENOENT") {
+        await fsPromises.writeFile(this.path, "", "utf-8");
+        return {} as Record<string, T>;
+      }
+      throw new Error(`Error reading and parsing file at path: ${this.path}`);
```
Comment (jacoblee93) [inline langchain/src/storage/file_system.ts:44]: Shouldn't we have something about the actual error in the error message?

File: langchain/src/storage/file_system.ts [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+  private async getParsedFile(key: string): Promise<Uint8Array | undefined> {
+    try {
+      const fileContent = await fsPromises.readFile(this.getFullPath(key));
+      return fileContent;
+      // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    } catch (e: any) {
+      // If the file does not exist, create it
+      if (("code" in e && e.code === "EISDIR") || e.code === "ENOENT") {
+        await fsPromises.writeFile(this.getFullPath(key), "");
+        return undefined;
```
Comment (jacoblee93) [inline langchain/src/storage/file_system.ts:33]: This will return `undefined` the first time and then empty string the second time won't it?

Comment (bracesproul) [reply langchain/src/storage/file_system.ts:33]: Ah yeah that's right since I removed the return undefined check

Comment (bracesproul) [reply langchain/src/storage/file_system.ts:36]: Fixed

File: langchain/src/storage/tests/file_system.test.ts [hunk 3]
```diff
@@ -0,0 +1,97 @@
+/* eslint-disable no-process-env */
+
+import { test } from "@jest/globals";
+import fsPromises from "fs/promises";
+import { LocalFileStore } from "../file_system.js";
+
+describe("LocalFileStore", () => {
+  const keys = ["key1", "key2"];
+  const rootPath = "./file_system_store_test";
```
Comment (jacoblee93) [inline langchain/src/storage/tests/file_system.test.ts:9]: Use `os.tempdir` here

https://github.com/langchain-ai/langchainjs/blob/4e35156fdf709ae22a85a79c1ce882b4e1ef4bd5/langchain/src/document_loaders/web/azure_blob_storage_file.ts#L64

**Review by jacoblee93 — CHANGES_REQUESTED** [review]

**Review by jacoblee93 — APPROVED** [review]

**Comment by jacoblee93** [comment]
Thanks!

**Comment by drewB** [comment]
May I ask what the purpose of this store is?  I tried using it with MultiVectorRetriever it is in incompatiable because it is declared as `BaseStore<string, Uint8Array>` and MultiVectorRetriever requires  `BaseStore<string, Document>`

**Comment by jacoblee93** [comment]
There is a wrapper class called `EncoderBackedStore` to make this usable there - we need docs on it though CC @bracesproul

**Comment by drewB** [comment]
Thanks.  So looking at that class, looks like I would want to do something like:

```typescript
const docstore = createDocumentStoreFromByteStore(new LocalFileStore({rootPath: localFileStoreDirectory}))

const retriever = new MultiVectorRetriever({
        vectorstore: vectorStore,
        docstore,
        idKey,
});
```

Correct?

**Comment by jacoblee93** [comment]
Yes that looks correct!

**Pull request merged**